### PR TITLE
Update input collections, rebased

### PR DESCRIPTION
**Pull request by gkaratha**
Hi all,

This is a rebased version of #15
Perhaps @amartelli  wants to look at this and please give a review / comment. Essentially instead of making a huge PR i broke it in two. Here is the first part - inputs
To summarize:
   - I rewrote the the cleaning with respect to triggering muon(s) for all inputs (muons, tracks, electrons) -> cleaner and faster
   - Made the electron merger to also cut and put BDT values so we need to loop only once instead of multiple times
   - All input objects that pass the cuts are labelled: Selected<NameOfObject>
   - Removed the two selections from track merger. If we are sure that K in tag objects are within certain DR from trg muon we can uncomment it
   - Fixed the addAsDaughter in composite candidate. Now the composite candidate has correct pT, eta, phi without "hackering it" as daughter. The candidate itself is added as additional Cand. I need to check that this works in later steps though.
    - Checked on 1k evts that results are fine
    - Measured time and size wrt to the flat ntuple
   - Fixed all the relevant cffs
   - Added comments questions in line so anyone can follow
If there are comments let me know

Best,
George

### CODE REVIEW

File: BParkingNano/plugins/TrackMerger.cc [hunk 1]
```diff
@@ -53,127 +57,120 @@ class TrackMerger : public edm::global::EDProducer<> {
...
+   GlobalPoint trgvtx;
+   for (const pat::Muon & mu: *trgMuons){
+    //remove tracks inside trg muons jet
+    if(reco::deltaR(trk, mu) < drTrg_Cleaning_ && drTrg_Cleaning_ >0) 
+      continue;
+    //if dz is negative it is deactivated
+    if((fabs(trk.vz() - mu.vz()) > dzTrg_cleaning_ && dzTrg_cleaning_ > 0))
+       continue;
+    skipTrack=false;
+    trgvtx=GlobalPoint(mu.vx(),mu.vy(),mu.vz());
```
Comment (sarafiorendi) [inline BParkingNano/plugins/TrackMerger.cc]: we should double check which is the vtx associated to the muon.
For example, if is it the PV[0] or a PV chosen based on the muon trajectory.

**Comment by mverzett** [comment]
@gkaratha the cfg for running fails with the following error:
```
cmsRun run_nano_cfg.py reportEvery=10 tag=$TAG
----- Begin Fatal Exception 08-Aug-2019 10:27:48 CEST-----------------------
An exception of category 'ConfigFileReadError' occurred while
   [0] Processing the python configuration file named run_nano_cfg.py
Exception Message:
python encountered the error: <type 'exceptions.NameError'>
global name 'tracksBParkSequence' is not defined
----- End Fatal Exception -------------------------------------------------
```

File: BParkingNano/plugins/ElectronMerger.cc [hunk 2]
```diff
@@ -38,93 +50,134 @@ class ElectronMerger : public edm::global::EDProducer<> {
...
+   bool skipEle=true;
+   for(const auto & trg : *trgMuon) {
+     if(reco::deltaR(ele, trg) < drTrg_cleaning_ && drTrg_cleaning_ > 0)
+        continue;
+     if(fabs(ele.vz() - trg.vz()) > dzTrg_cleaning_ && dzTrg_cleaning_ > 0)
+        continue;
+     skipEle=false;
+     break; // one trg muon to pass is enough :)
+   }
+   // we skip evts without trg muon
```
[the remark this answers is not in the document]
Comment (gkaratha) [reply BParkingNano/plugins/ElectronMerger.cc]: Hi Sara Thanks for the comments! The code does not discards electrons. If it finds at least A TRG Muon then it will use it to build the B. we only skip e in evts WITHOUT trg muon or if an e is outside of the cone of all trg muons

File: BParkingNano/plugins/ElectronMerger.cc [hunk 3]
```diff
@@ -38,93 +50,134 @@ class ElectronMerger : public edm::global::EDProducer<> {
...
+                                             ele.gsfTrack()->phiMode(),
+                                             ele.mass()    );
+     ele.setP4(p4);
+   }
+
+   //same cuts as in PF
+   if (ele.pt()<ptMin_) continue;
+   if (fabs(ele.eta())>etaMax_) continue;
+   // apply conversion veto?
+   if (!ele.passConversionVeto()) continue;
```
[the remark this answers is not in the document]
Comment (gkaratha) [reply BParkingNano/plugins/ElectronMerger.cc]: Every plot and efficiency that I did with low pT e had it already

[the remark this answers is not in the document]
Comment (sarafiorendi) [reply BParkingNano/plugins/ElectronMerger.cc]: yes I know, but do we know the effect of removing the cut? do we gain in efficiency?

File: BParkingNano/plugins/MuonTriggerSelector.cc [hunk 4]
```diff
@@ -138,18 +146,17 @@ void MuonTriggerSelector::produce(edm::Event& iEvent, const edm::EventSetup& iSe
...
     edm::Handle<std::vector<pat::Muon>> muons;
     iEvent.getByToken(muonSrc_, muons);
 
-    for(unsigned int iMuo=0; iMuo<muons->size(); ++iMuo){
-      const pat::Muon & muon = (*muons)[iMuo];
-
+    for(const pat::Muon & muon : *muons){
       //this is for triggering muon not really need to be configurable
+      unsigned int iMuo(&muon - &(muons->at(0)) );
       if(!(muon.isLooseMuon() && muon.isSoftMuon(PV))) continue;
```
[the remark this answers is not in the document]
Comment (gkaratha) [reply BParkingNano/plugins/MuonTriggerSelector.cc]: The idea is to remove as many objects as possible BEFORE going to Kin Fit. so it is PV. Furthermore if I am not mistakem soft muon ID was developed for B analysis. I think the main dilemma here should be: " use as PV the highest pT or trg muons?"

File: BParkingNano/plugins/TrackMerger.cc (hunk 1 above)
Comment (gkaratha) [reply BParkingNano/plugins/TrackMerger.cc]: Of course we should but according to the documentation is not necessarily the Highest pT one

File: BParkingNano/plugins/MuonTriggerSelector.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (sarafiorendi) [reply BParkingNano/plugins/MuonTriggerSelector.cc]: Soft ID was developed for B analysis but studying prompt Jpsi, therefore it was ok to use the PV. I'm just wondering if, given that our muons come from a secondary vtx, it makes sense to cut on the distance from the PV. On the other hand, it's true that the cuts are quite loose, and that for the triggering muon we don't want to reconstruct the full decay chain

File: BParkingNano/plugins/TrackMerger.cc (hunk 1 above)
Comment (sarafiorendi) [reply BParkingNano/plugins/TrackMerger.cc]: can you point me to the documentation you are referring to?

File: BParkingNano/plugins/TrackMerger.cc [hunk 5]
```diff
@@ -53,127 +57,120 @@ class TrackMerger : public edm::global::EDProducer<> {
...
+    break; // at least for one trg muon to pass this cuts
+   }
+   // if track is closer to at least a triggering muon keep it
+   if (skipTrack) continue;
+
+   // high purity requirment applied only in packedCands
+   unsigned int itrk=&trk-&totalTracks[0];
+   if( itrk < nTracks && !trk.trackHighPurity()) continue;
```
[the remark this answers is not in the document]
Comment (gkaratha) [reply BParkingNano/plugins/TrackMerger.cc]: well I do not think that there are much to gain here if we stick in MINIAOD:
    1) If i remember correctly from the track POG twiki, high purity is the one we should use for analysis
    2) MiniAOD primarily stores High purity tracks, so it does not make sense not to use the cut if we want to use miniaod
    3) PF cands are from tracks are high purity, Lost tracks can have a non HP contribution BUT they do pass other cuts. Lost Tracks ~20/evt, PF cands ~500/evt and from 20 most of those will get HP. So the contribution here is really small

File: BParkingNano/plugins/ElectronMerger.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (sarafiorendi) [reply BParkingNano/plugins/ElectronMerger.cc]: ah ok thanks! now I understand :) sorry for the noise

[the remark this answers is not in the document]
Comment (gkaratha) [reply BParkingNano/plugins/ElectronMerger.cc]: I should thank you for the comments and taking time to review it. I appreciated :)

**Comment by gkaratha** [comment]
Hi @mverzett , I am sorry for that... Probably messed up an name in cfg during the "poor mans" rebase. I will fix it now

File: BParkingNano/plugins/ElectronMerger.cc [hunk 6]
```diff
@@ -38,93 +50,134 @@ class ElectronMerger : public edm::global::EDProducer<> {
...
-	      out->begin(), out->end(), 
-	      [] (pat::Electron e1, pat::Electron e2) -> bool {return e1.pt() > e2.pt();}
+  
+  // -> changing order of loops ert Arabella's fix this without need for more vectors  
+ for(auto ele : *pf) {
+   //cuts
+   if (ele.pt()<ptMin_) continue;
+   if (fabs(ele.eta())>etaMax_) continue;
+   // apply conversion veto unless we want conversions
+   if (!ele.passConversionVeto()) continue;
```
Comment (mverzett) [inline BParkingNano/plugins/ElectronMerger.cc]: I would suggest to avoid the proliferation of configurable cut values by putting a `StringCutObjectSelector` like [here](https://github.com/mverzett/BAnalysisCode/blob/nano_v2/ParkingNtuples/plugins/DiLeptonBuilder.h#L27)

File: BParkingNano/plugins/ElectronMerger.cc (hunk 2 above)
Comment (mverzett) [inline BParkingNano/plugins/ElectronMerger.cc]: Given that this code is repeated exactly below, please make it an inline function, or filter the electrons after they are merged.

File: BParkingNano/plugins/MuonTriggerSelector.cc [hunk 7]
```diff
@@ -174,36 +181,36 @@ void MuonTriggerSelector::produce(edm::Event& iEvent, const edm::EventSetup& iSe
...
-	alreadySavedM[icount] = 1;
-	resultFilter->emplace_back(muon);
-      }
+    
+    // code simplified loop of trg inside
+      for(const pat::Muon &mu : *muons) {
+	//election cuts
+       if (mu.pt() < ptMin_) continue;
+       if (fabs(mu.eta()) > absEtaMax_) continue;
+       if (softMuonsOnly_ && !mu.isSoftMuon(PV)) continue;
```
Comment (mverzett) [inline BParkingNano/plugins/MuonTriggerSelector.cc]: Again, for the preselection, if possible, use the `StringCutObjectSelector`.

File: BParkingNano/plugins/TrackMerger.cc [hunk 8]
```diff
@@ -53,127 +57,120 @@ class TrackMerger : public edm::global::EDProducer<> {
...
-  int nLostTracks = lostTracks->size();      
-  int totalTracks = nTracks + nLostTracks;
+ // do we need all those ints?
+//  int nLostTracks = lostTracks->size();      
+//  int totalTracks = nTracks + nLostTracks;
+ //same as before comment out untill we are sure 
+//  std::unique_ptr<pat::CompositeCandidateCollection> outTag(new pat::CompositeCandidateCollection());
+
+//ok this was CompositeCandidateCollection 
+  std::unique_ptr<pat::CompositeCandidateCollection> tracks_out(new pat::CompositeCandidateCollection);
```
Comment (mverzett) [inline BParkingNano/plugins/TrackMerger.cc]: I am surprised it compiles, it should be `new pat::CompositeCandidateCollection()`

File: BParkingNano/plugins/TrackMerger.cc [hunk 9]
```diff
@@ -53,127 +57,120 @@ class TrackMerger : public edm::global::EDProducer<> {
...
+
+
+//correct logic but a bit convoluted -> changing to smthn simpler
+ std::vector<pat::PackedCandidate> totalTracks(*tracks);
+
+ totalTracks.insert(totalTracks.end(),lostTracks->begin(),lostTracks->end());
```
Comment (mverzett) [inline BParkingNano/plugins/TrackMerger.cc]: Add spaces after comas. Also, what is the need of copying them in the new track collection?

File: BParkingNano/plugins/TrackMerger.cc [hunk 10]
```diff
@@ -53,127 +57,120 @@ class TrackMerger : public edm::global::EDProducer<> {
...
+
+ 
+ // for loop is better to be range based - especially for large ensembles  
+ for( const pat::PackedCandidate & trk: totalTracks){
+
+   //arranging cuts for speed
+   if(!trk.hasTrackDetails()) continue;
+   if(abs(trk.pdgId()) != 211) continue; //do we want also to keep muons?
+   if(trk.pt() < trkPtCut_ ) continue;
+   if(fabs(trk.eta()) > trkEtaCut_) continue;
```
Comment (mverzett) [inline BParkingNano/plugins/TrackMerger.cc]: Again, I would suggest using `StringCutObjectSelector`

File: BParkingNano/plugins/TrackMerger.cc [hunk 11]
```diff
@@ -53,127 +57,120 @@ class TrackMerger : public edm::global::EDProducer<> {
...
+   pcand.setVertex(trk.vertex());
+   pcand.addUserInt("isPacked", (itrk < nTracks) ? 1 : 0);
+   pcand.addUserInt("isLostTrk", (itrk < nTracks) ? 0 : 1);      
+   pcand.addUserFloat("dxy", trk.dxy());
+   pcand.addUserFloat("dxyS", trk.dxy()/trk.dxyError());
+   pcand.addUserFloat("dz", trk.dz()); 
+   pcand.addUserFloat("dzS", trk.dz()/trk.dzError());
+   pcand.addUserFloat("DCASig", DCASig);
+   //adding the candidate in the composite stuff for fit (need to test)
+//   pcand.addUserCand("cand",trk.sourceCandidatePtr(0));
```
Comment (mverzett) [inline BParkingNano/plugins/TrackMerger.cc]: We should also store the pointer to the PackedCandidate, if we want to get the TransientTrack later on

File: BParkingNano/python/electronsBPark_cff.py [hunk 12]
```diff
@@ -77,7 +80,6 @@
         isLowPt = Var("userInt('isLowPt')",bool,doc="electron is LowPt candidate"),
         ptBiased = Var("userFloat('ptBiased')",float,doc="ptBiased from seed BDT 20 for pfEle"), 
         unBiased = Var("userFloat('unBiased')",float,doc="unBiased from seed BDT 20 for pfEle"), 
-        mvaId = Var("userFloat('mvaId')",float,doc="MVA ID for low pT, 20 for pfEle"),
```
Comment (mverzett) [inline BParkingNano/python/electronsBPark_cff.py]: This should be reintroduced

File: BParkingNano/python/electronsBPark_cff.py [hunk 13]
```diff
@@ -1,58 +1,61 @@
 import FWCore.ParameterSet.Config as cms
 from PhysicsTools.NanoAOD.common_cff import *
 
-from RecoEgamma.EgammaElectronProducers.lowPtGsfElectronID_cff import lowPtGsfElectronID
-lowPtGsfElectronLatestID = lowPtGsfElectronID.clone()
-lowPtGsfElectronLatestID.electrons = 'slimmedLowPtElectrons'
-lowPtGsfElectronLatestID.rho = 'fixedGridRhoFastjetAll'
-
```
Comment (mverzett) [inline BParkingNano/python/electronsBPark_cff.py]: Why are we not computing the electron ID any more?

**Comment by gkaratha** [comment]
Hi @mverzett ,

Thanks for the comments! For the general ones:
   1) ok 
   2) The reason is because I would like first to compare the code itself before I introduce cuts. Some cuts like Dz for example are not trivial so my idea was to take care of cpp, merging objects, b reco and then apply cuts and rerun. Again pt eta etc cuts are fine Dz is not simple though

File: BParkingNano/plugins/ElectronMerger.cc (hunk 6 above)
Comment (gkaratha) [reply BParkingNano/plugins/ElectronMerger.cc]: The problem with that approach is that it adds overhead time[*], and given that we want to run both on huge collections and to be fast at the same time applying everything in one loop is the most performing option. And since we are bound to have at least a loop for Dz cut and merging, is better to put everything there. Of course we can change it but that means that we are going to make more loops since Dz, which is necessary, cannot be apply in this Selector. Also cuts like min BDT in low pT e may become very tricky and cause one loop to evaluate, one to apply etc. Also it is a bit better in terms of bookkeeping since we now exactly which cuts will be used (they are configurable but IN our cfgs after all)


[*]What may cause a slight performance loss is the actual evaluation of expression. The parses, in fact, creates a composite structure in memory that invokes a virtual function call for every syntactical element in the expression.

Moreover, the passed object (track, in the example) has to be "converted" into a Reflex object in order to be evaluated.

File: BParkingNano/plugins/ElectronMerger.cc (hunk 2 above)
Comment (gkaratha) [reply BParkingNano/plugins/ElectronMerger.cc]: Yes I agree i will try inline function but inside the same loop. I would try to avoid another filter or another loop if that's possible, again for performance issues

File: BParkingNano/plugins/TrackMerger.cc (hunk 8 above)
Comment (gkaratha) [reply BParkingNano/plugins/TrackMerger.cc]: Actually me too... It is a typo. The fun fact is that compiles and result is good.

File: BParkingNano/plugins/TrackMerger.cc (hunk 9 above)
Comment (gkaratha) [reply BParkingNano/plugins/TrackMerger.cc]: Is there a way to do it without copying?

File: BParkingNano/plugins/TrackMerger.cc (hunk 11 above)
Comment (gkaratha) [reply BParkingNano/plugins/TrackMerger.cc]: Sure this i think is solved already ;)

File: BParkingNano/python/electronsBPark_cff.py (hunk 13 above)
Comment (gkaratha) [reply BParkingNano/python/electronsBPark_cff.py]: The above two comments: I do not know exactly how that happened but i will fix it. Thanks for noticing it

File: BParkingNano/plugins/ElectronMerger.cc (hunk 6 above)
Comment (mverzett) [reply BParkingNano/plugins/ElectronMerger.cc]: * The time overhead is really minimal, and no matter the size of the collection, the vertex creation will make it insignificant.
   *  No need for multiple loops, you can just 
      1. embed 
      2. evaluate cut parser
      3. check for DR and Dz

Comment (gkaratha) [reply BParkingNano/plugins/ElectronMerger.cc]: Good morning Mauro,

Well in order to embed (1 loop) evaluate kinematic cuts (2 loops) finally apply DR Dz cuts (3 loops) is essentially 3 times the time needed right? Well it would be indeed insignificant if we were to create every possible candidate. But we apply cuts so a really small part of all possible triplets (10^-4 ) will be promoted, while the loops on the inputs will be there for every event constantly.

Comment (mverzett) [reply BParkingNano/plugins/ElectronMerger.cc]: @gkaratha the number of loops would be unaffected, `StringCutObjectSelector ` is nothing but a function that evaluates to a bool that you can use in an if statement, you just replace the condition.

I forgot your comment regarding the reflex dictionaries. We (CMS) build reflex dictionaries for every object we store in the event (ROOT needs them) so they are already there, and anyhow those are done at compilation time, not at runtime: that's the reason why compiling the modules in DataFormats takes so long.

File: BParkingNano/python/electronsBPark_cff.py (hunk 13 above)
Comment (gkaratha) [reply BParkingNano/python/electronsBPark_cff.py]: Those are fixed now. Please check

File: BParkingNano/plugins/ElectronMerger.cc (hunk 6 above)
Comment (gkaratha) [reply BParkingNano/plugins/ElectronMerger.cc]: @mverzett , There is also the time reading and writing the new products which at least embeder and Dz cut will have (btw what is DR cut?). So even if selector makes no loops still the embeder and Dz will have. So everything will be 2 times slower, no matter how small that is. Now, no matter how small the total overhead per evt is, it is something that we will "pay" in every evt **for 10^10 evts** while fitter after cuts will be invoked for a small number of events. **So the end effect is not clear to me**. Also, at least in my point of view, is better to have a single producer per object, concentrating everything about the object. An additional point is that when we introduce muon+track and keep the same logic(embeder+selector+Dz) it will be more overhead in a code that is not very fast to begin with.  In any case i did not removed any of the files, just comment them out so they can be commented in :) 

Bottom line: The logic behind this change is to have the fastest code possible, and from the conversation I dont see any objection on that. Now if we want to have this or this+Selector or the embeder+selector+this is something we can discuss ;)
